src/controllers/helpers/pagination.rs

```diff
@@ -303,6 +315,61 @@ pub(crate) fn decode_seek<D: for<'a> Deserialize<'a>>(seek: &str) -> anyhow::Res
     Ok(decoded)
 }
 
+#[derive(Debug)]
+pub(crate) struct PaginatedQueryWithCountSubq<T, C> {
+    query: T,
+    count_query: C,
+    options: PaginationOptions,
+}
+
+impl<T, C> QueryId for PaginatedQueryWithCountSubq<T, C> {
+    const HAS_STATIC_QUERY_ID: bool = false;
+    type QueryId = ();
+}
+
+impl<
+        T: Query,
+        C: Query + QueryDsl + diesel::query_dsl::methods::SelectDsl<diesel::dsl::CountStar>,
+    > Query for PaginatedQueryWithCountSubq<T, C>
+{
+    type SqlType = (T::SqlType, BigInt);
+}
+
+impl<T, C, DB> RunQueryDsl<DB> for PaginatedQueryWithCountSubq<T, C> {}
+
+impl<T, C> QueryFragment<Pg> for PaginatedQueryWithCountSubq<T, C>
+where
+    T: QueryFragment<Pg>,
+    C: QueryFragment<Pg>,
+{
+    fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, Pg>) -> QueryResult<()> {
+        out.push_sql("SELECT *, (");
+        self.count_query.walk_ast(out.reborrow())?;
+        out.push_sql(") FROM (");
+        self.query.walk_ast(out.reborrow())?;
+        out.push_sql(") t LIMIT ");
+        out.push_bind_param::<BigInt, _>(&self.options.per_page)?;
+        if let Some(offset) = self.options.offset() {
+            out.push_sql(format!(" OFFSET {offset}").as_str());
```

Comment:
Thank you for addressing my concern in the issue about injection-- this line in particular, I had to look at surrounding code to convince myself that this wasn't a possible user injection location because `offset()` returns `Option<i64>` (as opposed to a user-provided string without any processing on our end, which could be Quite Bad Indeed).

Could you please add a comment to this line, much like the sort of comment that's good practice to leave when you write `unsafe` Rust code, that documents why this is safe? Something like:

```suggestion
            // Injection safety: `offset()` returns `Option<i64>`, so this interpolation is constrained to known
            // valid values and this is not vulnerable to user injection attacks.
            out.push_sql(format!(" OFFSET {offset}").as_str());
```

just to encode this assumption a bit more for the future? It would help me sleep better at night 😁 Thank you!

Reply:
I appreciate your suggestion, and you're absolutely right. Even I had to review the `push_sql` part multiple times to avoid unintended consequences.

We could even consider adding an `offset` field and using `push_bind_value` instead, which might reduce potential risks.

Reply:
```suggestion
            out.push_sql(" OFFSET ");
            out.push_bind_param::<BigInt, _>(&offset)?;
```

is there a reason why we're not using something like this?

looks like we're doing the same for the `per_page` value above, so I'm wondering if there is a reason for not doing the same here too :)

Reply:
My current implementation follows the existing pattern for consistency. However, if we want to apply the same approach to `offset` as with `per_page`, storing its value would be necessary to avoid potential lifetime issues.

Reply:
> storing its value would be necessary to avoid potential lifetime issues

I see. I guess the `format!()` resolves that by serializing it to a `String`. nevermind then :)